File: tests/algorithms/transpiler/FS/neural_network/input_data.fs
// Generated 2025-08-22 13:05 +0700

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _idx (arr:'a array) (i:int) : 'a =
    if not (obj.ReferenceEquals(arr, null)) && i >= 0 && i < arr.Length then arr.[i] else Unchecked.defaultof<'a>
let rec _str v =
    match box v with
    | :? float as f -> sprintf "%.10g" f
    | _ ->
        let s = sprintf "%A" v
        s.Replace("[|", "[")
         .Replace("|]", "]")
         .Replace("; ", " ")
         .Replace(";", "")
         .Replace("\"", "")
type DataSet = {
    mutable images: int array array
    mutable labels: int array array
    mutable _num_examples: int
    mutable _index_in_epoch: int
    mutable _epochs_completed: int
}
type Datasets = {
    mutable _train: DataSet
    mutable _validation: DataSet
    mutable _test_ds: DataSet
}
type BatchResult = {
    mutable _dataset: DataSet
    mutable images: int array array
    mutable labels: int array array
}
let rec dense_to_one_hot (labels: int array) (num_classes: int) =
    let mutable __ret : int array array = Unchecked.defaultof<int array array>
    let mutable labels = labels
    let mutable num_classes = num_classes
    try
        let mutable result: int array array = Array.empty<int array>
        let mutable i: int = 0
        while i < (Seq.length (labels)) do
            let mutable row: int array = Array.empty<int>
            let mutable j: int = 0
            while j < num_classes do
                if j = (_idx labels (int i)) then
                    row <- Array.append row [|1|]
                else
                    row <- Array.append row [|0|]
                j <- j + 1
            result <- Array.append result [|row|]
            i <- i + 1
        __ret <- result
        raise Return
        __ret
    with
        | Return -> __ret
and new_dataset (images: int array array) (labels: int array array) =
    let mutable __ret : DataSet = Unchecked.defaultof<DataSet>
    let mutable images = images
    let mutable labels = labels
    try
        __ret <- { images = images; labels = labels; _num_examples = Seq.length (images); _index_in_epoch = 0; _epochs_completed = 0 }
        raise Return
        __ret
    with
        | Return -> __ret
and next_batch (ds: DataSet) (batch_size: int) =
    let mutable __ret : BatchResult = Unchecked.defaultof<BatchResult>
    let mutable ds = ds
    let mutable batch_size = batch_size
    try
        let start: int = ds._index_in_epoch
        if (start + batch_size) > (ds._num_examples) then
            let rest: int = (ds._num_examples) - start
            let images_rest: int array array = Array.sub ds.images start ((ds._num_examples) - start)
            let labels_rest: int array array = Array.sub ds.labels start ((ds._num_examples) - start)
            let new_index: int = batch_size - rest
            let images_new: int array array = Array.sub ds.images 0 (new_index - 0)
            let labels_new: int array array = Array.sub ds.labels 0 (new_index - 0)
            let batch_images: int array array = Array.append (images_rest) (images_new)
            let batch_labels: int array array = Array.append (labels_rest) (labels_new)
            let new_ds: DataSet = { images = ds.images; labels = ds.labels; _num_examples = ds._num_examples; _index_in_epoch = new_index; _epochs_completed = (ds._epochs_completed) + 1 }
            __ret <- { _dataset = new_ds; images = batch_images; labels = batch_labels }
            raise Return
        else
            let ``end``: int = start + batch_size
            let batch_images: int array array = Array.sub ds.images start (``end`` - start)
            let batch_labels: int array array = Array.sub ds.labels start (``end`` - start)
            let new_ds: DataSet = { images = ds.images; labels = ds.labels; _num_examples = ds._num_examples; _index_in_epoch = ``end``; _epochs_completed = ds._epochs_completed }
            __ret <- { _dataset = new_ds; images = batch_images; labels = batch_labels }
            raise Return
        __ret
    with
        | Return -> __ret
and read_data_sets (train_images: int array array) (train_labels_raw: int array) (test_images: int array array) (test_labels_raw: int array) (validation_size: int) (num_classes: int) =
    let mutable __ret : Datasets = Unchecked.defaultof<Datasets>
    let mutable train_images = train_images
    let mutable train_labels_raw = train_labels_raw
    let mutable test_images = test_images
    let mutable test_labels_raw = test_labels_raw
    let mutable validation_size = validation_size
    let mutable num_classes = num_classes
    try
        let train_labels: int array array = dense_to_one_hot (train_labels_raw) (num_classes)
        let test_labels: int array array = dense_to_one_hot (test_labels_raw) (num_classes)
        let validation_images: int array array = Array.sub train_images 0 (validation_size - 0)
        let validation_labels: int array array = Array.sub train_labels 0 (validation_size - 0)
        let train_images_rest: int array array = Array.sub train_images validation_size ((Seq.length (train_images)) - validation_size)
        let train_labels_rest: int array array = Array.sub train_labels validation_size ((Seq.length (train_labels)) - validation_size)
        let _train: DataSet = new_dataset (train_images_rest) (train_labels_rest)
        let _validation: DataSet = new_dataset (validation_images) (validation_labels)
        let testset: DataSet = new_dataset (test_images) (test_labels)
        __ret <- { _train = _train; _validation = _validation; _test_ds = testset }
        raise Return
        __ret
    with
        | Return -> __ret
and main () =
    let mutable __ret : obj = Unchecked.defaultof<obj>
    try
        let __bench_start = _now()
        let __mem_start = System.GC.GetTotalMemory(true)
        let train_images: int array array = [|[|0; 1|]; [|1; 2|]; [|2; 3|]; [|3; 4|]; [|4; 5|]|]
        let train_labels_raw: int array = unbox<int array> [|0; 1; 2; 3; 4|]
        let test_images: int array array = [|[|5; 6|]; [|6; 7|]|]
        let test_labels_raw: int array = unbox<int array> [|5; 6|]
        let data: Datasets = read_data_sets (train_images) (train_labels_raw) (test_images) (test_labels_raw) (2) (10)
        let mutable ds: DataSet = data._train
        let mutable res: BatchResult = next_batch (ds) (2)
        ds <- res._dataset
        ignore (printfn "%s" (_str (res.images)))
        ignore (printfn "%s" (_str (res.labels)))
        res <- next_batch (ds) (2)
        ds <- res._dataset
        ignore (printfn "%s" (_str (res.images)))
        ignore (printfn "%s" (_str (res.labels)))
        res <- next_batch (ds) (2)
        ds <- res._dataset
        ignore (printfn "%s" (_str (res.images)))
        ignore (printfn "%s" (_str (res.labels)))
        let __bench_end = _now()
        let __mem_end = System.GC.GetTotalMemory(true)
        printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

        __ret
    with
        | Return -> __ret
ignore (main())

File: tests/algorithms/x/Java/neural_network/input_data.java
public class Main {
    static class DataSet {
        java.math.BigInteger[][] images;
        java.math.BigInteger[][] labels;
        java.math.BigInteger num_examples;
        java.math.BigInteger index_in_epoch;
        java.math.BigInteger epochs_completed;
        DataSet(java.math.BigInteger[][] images, java.math.BigInteger[][] labels, java.math.BigInteger num_examples, java.math.BigInteger index_in_epoch, java.math.BigInteger epochs_completed) {
            this.images = images;
            this.labels = labels;
            this.num_examples = num_examples;
            this.index_in_epoch = index_in_epoch;
            this.epochs_completed = epochs_completed;
        }
        DataSet() {}
        @Override public String toString() {
            return String.format("{'images': %s, 'labels': %s, 'num_examples': %s, 'index_in_epoch': %s, 'epochs_completed': %s}", String.valueOf(images), String.valueOf(labels), String.valueOf(num_examples), String.valueOf(index_in_epoch), String.valueOf(epochs_completed));
        }
    }

    static class Datasets {
        DataSet train;
        DataSet validation;
        DataSet test_ds;
        Datasets(DataSet train, DataSet validation, DataSet test_ds) {
            this.train = train;
            this.validation = validation;
            this.test_ds = test_ds;
        }
        Datasets() {}
        @Override public String toString() {
            return String.format("{'train': %s, 'validation': %s, 'test_ds': %s}", String.valueOf(train), String.valueOf(validation), String.valueOf(test_ds));
        }
    }

    static class BatchResult {
        DataSet dataset;
        java.math.BigInteger[][] images;
        java.math.BigInteger[][] labels;
        BatchResult(DataSet dataset, java.math.BigInteger[][] images, java.math.BigInteger[][] labels) {
            this.dataset = dataset;
            this.images = images;
            this.labels = labels;
        }
        BatchResult() {}
        @Override public String toString() {
            return String.format("{'dataset': %s, 'images': %s, 'labels': %s}", String.valueOf(dataset), String.valueOf(images), String.valueOf(labels));
        }
    }


    static java.math.BigInteger[][] dense_to_one_hot(java.math.BigInteger[] labels, java.math.BigInteger num_classes) {
        java.math.BigInteger[][] result = ((java.math.BigInteger[][])(new java.math.BigInteger[][]{}));
        java.math.BigInteger i_1 = java.math.BigInteger.valueOf(0);
        while (i_1.compareTo(new java.math.BigInteger(String.valueOf(labels.length))) < 0) {
            java.math.BigInteger[] row_1 = ((java.math.BigInteger[])(new java.math.BigInteger[]{}));
            java.math.BigInteger j_1 = java.math.BigInteger.valueOf(0);
            while (j_1.compareTo(num_classes) < 0) {
                if (j_1.compareTo(labels[_idx((labels).length, ((java.math.BigInteger)(i_1)).longValue())]) == 0) {
                    row_1 = ((java.math.BigInteger[])(java.util.stream.Stream.concat(java.util.Arrays.stream(row_1), java.util.stream.Stream.of(java.math.BigInteger.valueOf(1))).toArray(java.math.BigInteger[]::new)));
                } else {
                    row_1 = ((java.math.BigInteger[])(java.util.stream.Stream.concat(java.util.Arrays.stream(row_1), java.util.stream.Stream.of(java.math.BigInteger.valueOf(0))).toArray(java.math.BigInteger[]::new)));
                }
                j_1 = j_1.add(java.math.BigInteger.valueOf(1));
            }
            result = ((java.math.BigInteger[][])(java.util.stream.Stream.concat(java.util.Arrays.stream(result), java.util.stream.Stream.of(new java.math.BigInteger[][]{((java.math.BigInteger[])(row_1))})).toArray(java.math.BigInteger[][]::new)));
            i_1 = i_1.add(java.math.BigInteger.valueOf(1));
        }
        return ((java.math.BigInteger[][])(result));
    }

    static DataSet new_dataset(java.math.BigInteger[][] images, java.math.BigInteger[][] labels) {
        return new DataSet(((java.math.BigInteger[][])(images)), ((java.math.BigInteger[][])(labels)), new java.math.BigInteger(String.valueOf(images.length)), java.math.BigInteger.valueOf(0), java.math.BigInteger.valueOf(0));
    }

    static BatchResult next_batch(DataSet ds, java.math.BigInteger batch_size) {
        java.math.BigInteger start = ds.index_in_epoch;
        if (start.add(batch_size).compareTo(ds.num_examples) > 0) {
            java.math.BigInteger rest_1 = ds.num_examples.subtract(start);
            java.math.BigInteger[][] images_rest_1 = ((java.math.BigInteger[][])(java.util.Arrays.copyOfRange(ds.images, (int)(((java.math.BigInteger)(start)).longValue()), (int)(((java.math.BigInteger)(ds.num_examples)).longValue()))));
            java.math.BigInteger[][] labels_rest_1 = ((java.math.BigInteger[][])(java.util.Arrays.copyOfRange(ds.labels, (int)(((java.math.BigInteger)(start)).longValue()), (int)(((java.math.BigInteger)(ds.num_examples)).longValue()))));
            java.math.BigInteger new_index_1 = batch_size.subtract(rest_1);
            java.math.BigInteger[][] images_new_1 = ((java.math.BigInteger[][])(java.util.Arrays.copyOfRange(ds.images, (int)(0L), (int)(((java.math.BigInteger)(new_index_1)).longValue()))));
            java.math.BigInteger[][] labels_new_1 = ((java.math.BigInteger[][])(java.util.Arrays.copyOfRange(ds.labels, (int)(0L), (int)(((java.math.BigInteger)(new_index_1)).longValue()))));
            java.math.BigInteger[][] batch_images_2 = ((java.math.BigInteger[][])(concat(images_rest_1, images_new_1)));
            java.math.BigInteger[][] batch_labels_2 = ((java.math.BigInteger[][])(concat(labels_rest_1, labels_new_1)));
            DataSet new_ds_2 = new DataSet(((java.math.BigInteger[][])(ds.images)), ((java.math.BigInteger[][])(ds.labels)), ds.num_examples, new_index_1, ds.epochs_completed.add(java.math.BigInteger.valueOf(1)));
            return new BatchResult(new_ds_2, ((java.math.BigInteger[][])(batch_images_2)), ((java.math.BigInteger[][])(batch_labels_2)));
        } else {
            java.math.BigInteger end_1 = start.add(batch_size);
            java.math.BigInteger[][] batch_images_3 = ((java.math.BigInteger[][])(java.util.Arrays.copyOfRange(ds.images, (int)(((java.math.BigInteger)(start)).longValue()), (int)(((java.math.BigInteger)(end_1)).longValue()))));
            java.math.BigInteger[][] batch_labels_3 = ((java.math.BigInteger[][])(java.util.Arrays.copyOfRange(ds.labels, (int)(((java.math.BigInteger)(start)).longValue()), (int)(((java.math.BigInteger)(end_1)).longValue()))));
            DataSet new_ds_3 = new DataSet(((java.math.BigInteger[][])(ds.images)), ((java.math.BigInteger[][])(ds.labels)), ds.num_examples, end_1, ds.epochs_completed);
            return new BatchResult(new_ds_3, ((java.math.BigInteger[][])(batch_images_3)), ((java.math.BigInteger[][])(batch_labels_3)));
        }
    }

    static Datasets read_data_sets(java.math.BigInteger[][] train_images, java.math.BigInteger[] train_labels_raw, java.math.BigInteger[][] test_images, java.math.BigInteger[] test_labels_raw, java.math.BigInteger validation_size, java.math.BigInteger num_classes) {
        java.math.BigInteger[][] train_labels = ((java.math.BigInteger[][])(dense_to_one_hot(((java.math.BigInteger[])(train_labels_raw)), num_classes)));
        java.math.BigInteger[][] test_labels_1 = ((java.math.BigInteger[][])(dense_to_one_hot(((java.math.BigInteger[])(test_labels_raw)), num_classes)));
        java.math.BigInteger[][] validation_images_1 = ((java.math.BigInteger[][])(java.util.Arrays.copyOfRange(train_images, (int)(0L), (int)(((java.math.BigInteger)(validation_size)).longValue()))));
        java.math.BigInteger[][] validation_labels_1 = ((java.math.BigInteger[][])(java.util.Arrays.copyOfRange(train_labels, (int)(0L), (int)(((java.math.BigInteger)(validation_size)).longValue()))));
        java.math.BigInteger[][] train_images_rest_1 = ((java.math.BigInteger[][])(java.util.Arrays.copyOfRange(train_images, (int)(((java.math.BigInteger)(validation_size)).longValue()), (int)((long)(train_images.length)))));
        java.math.BigInteger[][] train_labels_rest_1 = ((java.math.BigInteger[][])(java.util.Arrays.copyOfRange(train_labels, (int)(((java.math.BigInteger)(validation_size)).longValue()), (int)((long)(train_labels.length)))));
        DataSet train_1 = new_dataset(((java.math.BigInteger[][])(train_images_rest_1)), ((java.math.BigInteger[][])(train_labels_rest_1)));
        DataSet validation_1 = new_dataset(((java.math.BigInteger[][])(validation_images_1)), ((java.math.BigInteger[][])(validation_labels_1)));
        DataSet testset_1 = new_dataset(((java.math.BigInteger[][])(test_images)), ((java.math.BigInteger[][])(test_labels_1)));
        return new Datasets(train_1, validation_1, testset_1);
    }

    static void main() {
        java.math.BigInteger[][] train_images = ((java.math.BigInteger[][])(new java.math.BigInteger[][]{((java.math.BigInteger[])(new java.math.BigInteger[]{java.math.BigInteger.valueOf(0), java.math.BigInteger.valueOf(1)})), ((java.math.BigInteger[])(new java.math.BigInteger[]{java.math.BigInteger.valueOf(1), java.math.BigInteger.valueOf(2)})), ((java.math.BigInteger[])(new java.math.BigInteger[]{java.math.BigInteger.valueOf(2), java.math.BigInteger.valueOf(3)})), ((java.math.BigInteger[])(new java.math.BigInteger[]{java.math.BigInteger.valueOf(3), java.math.BigInteger.valueOf(4)})), ((java.math.BigInteger[])(new java.math.BigInteger[]{java.math.BigInteger.valueOf(4), java.math.BigInteger.valueOf(5)}))}));
        java.math.BigInteger[] train_labels_raw_1 = ((java.math.BigInteger[])(new java.math.BigInteger[]{java.math.BigInteger.valueOf(0), java.math.BigInteger.valueOf(1), java.math.BigInteger.valueOf(2), java.math.BigInteger.valueOf(3), java.math.BigInteger.valueOf(4)}));
        java.math.BigInteger[][] test_images_1 = ((java.math.BigInteger[][])(new java.math.BigInteger[][]{((java.math.BigInteger[])(new java.math.BigInteger[]{java.math.BigInteger.valueOf(5), java.math.BigInteger.valueOf(6)})), ((java.math.BigInteger[])(new java.math.BigInteger[]{java.math.BigInteger.valueOf(6), java.math.BigInteger.valueOf(7)}))}));
        java.math.BigInteger[] test_labels_raw_1 = ((java.math.BigInteger[])(new java.math.BigInteger[]{java.math.BigInteger.valueOf(5), java.math.BigInteger.valueOf(6)}));
        Datasets data_1 = read_data_sets(((java.math.BigInteger[][])(train_images)), ((java.math.BigInteger[])(train_labels_raw_1)), ((java.math.BigInteger[][])(test_images_1)), ((java.math.BigInteger[])(test_labels_raw_1)), java.math.BigInteger.valueOf(2), java.math.BigInteger.valueOf(10));
        DataSet ds_1 = data_1.train;
        BatchResult res_1 = next_batch(ds_1, java.math.BigInteger.valueOf(2));
        ds_1 = res_1.dataset;
        System.out.println(_p(res_1.images));
        System.out.println(_p(res_1.labels));
        res_1 = next_batch(ds_1, java.math.BigInteger.valueOf(2));
        ds_1 = res_1.dataset;
        System.out.println(_p(res_1.images));
        System.out.println(_p(res_1.labels));
        res_1 = next_batch(ds_1, java.math.BigInteger.valueOf(2));
        ds_1 = res_1.dataset;
        System.out.println(_p(res_1.images));
        System.out.println(_p(res_1.labels));
    }
    public static void main(String[] args) {
        {
            long _benchStart = _now();
            long _benchMem = _mem();
            main();
            long _benchDuration = _now() - _benchStart;
            long _benchMemory = _mem() - _benchMem;
            System.out.println("{\"duration_us\": " + _benchDuration + ", \"memory_bytes\": " + _benchMemory + ", \"name\": \"main\"}");
            return;
        }
    }

    static boolean _nowSeeded = false;
    static int _nowSeed;
    static int _now() {
        if (!_nowSeeded) {
            String s = System.getenv("MOCHI_NOW_SEED");
            if (s != null && !s.isEmpty()) {
                try { _nowSeed = Integer.parseInt(s); _nowSeeded = true; } catch (Exception e) {}
            }
        }
        if (_nowSeeded) {
            _nowSeed = (int)((_nowSeed * 1664525L + 1013904223) % 2147483647);
            return _nowSeed;
        }
        return (int)(System.nanoTime() / 1000);
    }

    static long _mem() {
        Runtime rt = Runtime.getRuntime();
        rt.gc();
        return rt.totalMemory() - rt.freeMemory();
    }

    static Object concat(Object a, Object b) {
        int len1 = java.lang.reflect.Array.getLength(a);
        int len2 = java.lang.reflect.Array.getLength(b);
        Object out = java.lang.reflect.Array.newInstance(a.getClass().getComponentType(), len1 + len2);
        System.arraycopy(a, 0, out, 0, len1);
        System.arraycopy(b, 0, out, len1, len2);
        return out;
    }

    static String _p(Object v) {
        if (v == null) return "<nil>";
        if (v.getClass().isArray()) {
            if (v instanceof int[]) return java.util.Arrays.toString((int[]) v);
            if (v instanceof long[]) return java.util.Arrays.toString((long[]) v);
            if (v instanceof double[]) return java.util.Arrays.toString((double[]) v);
            if (v instanceof boolean[]) return java.util.Arrays.toString((boolean[]) v);
            if (v instanceof byte[]) return java.util.Arrays.toString((byte[]) v);
            if (v instanceof char[]) return java.util.Arrays.toString((char[]) v);
            if (v instanceof short[]) return java.util.Arrays.toString((short[]) v);
            if (v instanceof float[]) return java.util.Arrays.toString((float[]) v);
            return java.util.Arrays.deepToString((Object[]) v);
        }
        if (v instanceof java.util.Map<?, ?>) {
            StringBuilder sb = new StringBuilder("{");
            boolean first = true;
            for (java.util.Map.Entry<?, ?> e : ((java.util.Map<?, ?>) v).entrySet()) {
                if (!first) sb.append(", ");
                sb.append(_p(e.getKey()));
                sb.append("=");
                sb.append(_p(e.getValue()));
                first = false;
            }
            sb.append("}");
            return sb.toString();
        }
        if (v instanceof java.util.List<?>) {
            StringBuilder sb = new StringBuilder("[");
            boolean first = true;
            for (Object e : (java.util.List<?>) v) {
                if (!first) sb.append(", ");
                sb.append(_p(e));
                first = false;
            }
            sb.append("]");
            return sb.toString();
        }
        if (v instanceof Double || v instanceof Float) {
            double d = ((Number) v).doubleValue();
            return String.valueOf(d);
        }
        return String.valueOf(v);
    }

    static int _idx(int len, long i) {
        return (int)(i < 0 ? len + i : i);
    }
}
